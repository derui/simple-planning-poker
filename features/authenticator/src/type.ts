import { User } from "@spp/shared-domain";

/**
 * An authenticator for user signin/up
 */
export interface Authenticator {
  /**
   * Get logined user id or not
   */
  currentUserIdIfExists(): Promise<User.Id | undefined>;

  /**
   * Sign in with email and password
   */
  signIn(email: string, password: string): Promise<User.Id | undefined>;

  /**
   * Sign up with name, email and password
   */
  signUp(name: string, email: string, password: string): Promise<User.Id | undefined>;
}
