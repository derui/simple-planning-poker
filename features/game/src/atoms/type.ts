/**
 * status of Vote starting
 */
export enum VoteStartingStatus {
  Started = "started",
  Starting = "starting",
}

/**
 * Status enum of game creating
 */
export enum CreateGameStatus {
  /**
   * Completed state is completed to create a game
   */
  Completed = "completed",

  /**
   * Failed state is failed to create a game
   */
  Failed = "failed",

  /**
   * Waiting state is waiting to finish creating a game
   */
  Waiting = "waiting",
}
