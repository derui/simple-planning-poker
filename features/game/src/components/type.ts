/**
 * Voter type
 */
export enum VoterMode {
  /**
   * Mapping of VoterType of domain
   */
  Normal = "normal",

  /**
   * Mapping of VoterType of domain
   */
  Inspector = "inspector",
}
