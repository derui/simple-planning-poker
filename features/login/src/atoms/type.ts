/**
 * Authentication status. Not same as login status.
 */
export enum AuthStatus {
  Checking = "checking",
  NotAuthenticated = "notAuthenticated",
  Authenticated = "authenticated",
}

/**
 * Login status
 */
export enum LoginStatus {
  Logined = "logined",
  NotLogined = "notLogined",
  Doing = "doing",
}
