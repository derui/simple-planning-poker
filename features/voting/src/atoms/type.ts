import { UserRole } from "../components/types.js";
import { EstimationDto } from "./dto.js";

/**
 * Hook to get common attributes at polling place
 */
export type PollingPlace = {
  /**
   * ID of current voting
   */
  id: string;

  /**
   * Estimations in voting.
   */
  estimations: EstimationDto[];

  /**
   * inspectors in voting
   */
  inspectors: EstimationDto[];

  /**
   * theme of current joining voting
   */
  theme: string;

  /**
   * role of login user.
   */
  userRole: UserRole;

  /**
   * applicable points in this voting
   */
  points: string[];
};
