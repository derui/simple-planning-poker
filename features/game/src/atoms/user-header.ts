import { UseLoginUser } from "@spp/feature-login";
import { User, UserRepository, VoterType } from "@spp/shared-domain";
import { ChangeDefaultVoterTypeUseCase, ChangeUserNameUseCase } from "@spp/shared-use-case";
import { atom, useAtom, useAtomValue } from "jotai";
import { useEffect } from "react";
import { VoterMode } from "../components/type.js";
import { toUserDto, UserDto } from "./dto.js";

type UserHeaderStatus = "loading" | "loaded" | "editing" | "edited";
const statusAtom = atom<UserHeaderStatus>("loading");
const loginUserAtom = atom<User.T | undefined>(undefined);
const voterModeAtom = atom<VoterMode | undefined>((get) => {
  const user = get(loginUserAtom);

  if (!user) {
    return;
  }
  const voterType = user.defaultVoterType;
  if (voterType === VoterType.Normal) {
    return VoterMode.Normal;
  }
  if (voterType === VoterType.Inspector) {
    return VoterMode.Inspector;
  }
});

/**
 * Hook definition to list game
 */
export type UseUserHeader = () => {
  readonly status: UserHeaderStatus;

  readonly loginUser?: UserDto;

  /**
   * current voter mode of current user
   */
  readonly voterMode?: VoterMode;

  /**
   * Edit name of current user
   */
  readonly editName: (newName: string) => void;

  /**
   *  Change default voter mode of current user
   */
  readonly changeDefaultVoterMode: (voterMode: VoterMode) => void;
};

/**
 * Create hook implementation of `UseListGame`
 */
export const createUseUserHeader = function createUseUserHeader({
  useLoginUser,
  userRepository,
  changeUserNameUseCase,
  changeDefaultVoterModeUseCase,
}: {
  useLoginUser: UseLoginUser;
  userRepository: UserRepository.T;
  changeUserNameUseCase: ChangeUserNameUseCase;
  changeDefaultVoterModeUseCase: ChangeDefaultVoterTypeUseCase;
}): UseUserHeader {
  return () => {
    const [status, setStatus] = useAtom(statusAtom);
    const [loginUser, setLoginUser] = useAtom(loginUserAtom);
    const { userId } = useLoginUser();
    const voterMode = useAtomValue(voterModeAtom);

    useEffect(() => {
      setStatus("loading");

      if (userId) {
        userRepository
          .findBy(userId)
          .then((user) => {
            setLoginUser(user);
          })
          .catch(() => setLoginUser(undefined))
          .finally(() => setStatus("loaded"));
      }
    }, [userId, setStatus, setLoginUser]);

    return {
      status: status,
      loginUser: loginUser ? toUserDto(loginUser) : undefined,
      voterMode,
      editName: (newName: string) => {
        if (!loginUser) {
          return;
        }

        setStatus("editing");
        changeUserNameUseCase({
          userId: loginUser.id,
          name: newName,
        })
          .then((ret) => {
            switch (ret.kind) {
              case "success":
                setLoginUser(ret.user);
                break;
              default:
                setLoginUser(undefined);
                break;
            }
          })
          .finally(() => {
            setStatus("edited");
          });
      },

      changeDefaultVoterMode: (newMode: VoterMode): void => {
        if (!loginUser) {
          return;
        }

        const voterType = newMode == VoterMode.Inspector ? VoterType.Inspector : VoterType.Normal;

        setStatus("editing");
        changeDefaultVoterModeUseCase({
          userId: loginUser.id,
          voterType,
        })
          .then((ret) => {
            switch (ret.kind) {
              case "success":
                setLoginUser(ret.user);
                break;
              default:
                setLoginUser(undefined);
                break;
            }
          })
          .finally(() => {
            setStatus("edited");
          });
      },
    };
  };
};
