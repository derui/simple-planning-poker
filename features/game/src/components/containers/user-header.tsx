import { useEffect, useState } from "react";
import { hooks } from "../../hooks/facade.js";
import { UserHeaderLayout } from "./user-header.layout.js";

/**
 * Container for the user header.
 */
export const UserHeader = function UserHeader(): JSX.Element {
  const { loginUser, loading: status, changeDefaultVoterMode, editName } = hooks.useUserHeader();
  const [mode, setMode] = useState<"edit" | "view">("view");

  useEffect(() => {
    if (status == "edited") {
      setMode("view");
    }
  }, [status]);

  return (
    <UserHeaderLayout
      user={loginUser}
      mode={mode}
      onSubmitEdit={editName}
      onChangeDefaultVoterMode={changeDefaultVoterMode}
      onRequestUserNameEdit={() => setMode("edit")}
      onCancelUserNameEdit={() => setMode("view")}
    />
  );
};
