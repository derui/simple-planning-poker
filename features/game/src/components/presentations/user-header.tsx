import { Variant } from "@spp/shared-color-variant";
import { Icon } from "@spp/ui-icon";
import { VoterMode } from "../type.js";
import * as styles from "./user-header.css.js";

export interface Props {
  /**
   * The user's name.
   */
  readonly userName: string;

  /**
   * The user's default voter mode.
   */
  readonly defaultVoterMode: VoterMode;

  /**
   * The user's default voter mode.
   */
  readonly onChangeDefaultVoterMode?: (next: VoterMode) => void;

  /**
   * The user' name change request callback.
   */
  readonly onRequestUserNameEdit?: () => void;
}

export const UserHeader = function UserHeader({
  userName,
  defaultVoterMode,
  onChangeDefaultVoterMode,
  onRequestUserNameEdit,
}: Props): JSX.Element {
  const HeaderIcon = defaultVoterMode == VoterMode.Inspector ? Icon.Eye : Icon.User;

  const handleChangeDefaultVoterMode = () => {
    const next = defaultVoterMode == VoterMode.Inspector ? VoterMode.Normal : VoterMode.Inspector;

    onChangeDefaultVoterMode?.(next);
  };

  const handleRequestUserNameEdit = (event: React.MouseEvent) => {
    event.stopPropagation();
    event.preventDefault();

    onRequestUserNameEdit?.();
  };

  return (
    <div className={styles.root}>
      <button className={styles.voterMode} onClick={handleChangeDefaultVoterMode} aria-label="Toggle voter mode">
        <HeaderIcon size="m" variant={Variant.orange} />
      </button>
      <div className={styles.userName.container}>
        <span className={styles.userName.name}>{userName}</span>
        <span className={styles.userName.icon}>
          <button className={styles.userName.icon} onClick={handleRequestUserNameEdit} aria-label="Edit user name">
            <Icon.Pencil size="m" variant={Variant.teal} />
          </button>
        </span>
      </div>
    </div>
  );
};
