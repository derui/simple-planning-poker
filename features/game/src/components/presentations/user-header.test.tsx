import { cleanup, render } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import sinon from "sinon";
import { afterEach, expect, test } from "vitest";
import { VoterMode } from "../type.js";
import { UserHeader } from "./user-header.js";

afterEach(cleanup);

test("should be able to render", async () => {
  const ret = render(<UserHeader userName="name" defaultVoterMode={VoterMode.Normal} />);

  expect(ret.container).toMatchSnapshot();
});

test("should call callback when mode icon clicked", async () => {
  // Arrange
  const callback = sinon.fake();
  const ret = render(
    <UserHeader userName="name" defaultVoterMode={VoterMode.Normal} onChangeDefaultVoterMode={callback} />
  );

  // Act
  await userEvent.click(ret.getByRole("button", { name: "Toggle voter mode" }));

  // Assert
  expect(callback.calledWith(VoterMode.Inspector)).toBeTruthy();
});

test("should call callback when user name change button clicked", async () => {
  // Arrange
  const callback = sinon.fake();
  const ret = render(
    <UserHeader userName="name" defaultVoterMode={VoterMode.Normal} onRequestUserNameEdit={callback} />
  );

  // Act
  await userEvent.click(ret.getByRole("button", { name: "Edit user name" }));

  // Assert
  expect(callback.calledWith()).toBeTruthy();
});
