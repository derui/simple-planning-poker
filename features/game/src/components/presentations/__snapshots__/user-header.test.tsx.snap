// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to render 1`] = `
<div>
  <div
    class="user-header_root__44nqd40 style_transition_all__192bcwg41"
  >
    <button
      aria-label="Toggle voter mode"
      class="user-header_voterMode__44nqd44 style_enabledIconButtonVariants_orange__92u52z11 style_enabledIconBase__92u52z5 style_iconBase__92u52z1 style_transition_all__192bcwg41"
    >
      <span
        class="user_medium_orange__icoifyd style_iconMedium__p90ga82 style_iconBase__p90ga80"
      />
    </button>
    <div
      class="user-header_container__44nqd41 style_transition_all__192bcwg41"
    >
      <span
        class="user-header_name__44nqd42"
      >
        name
      </span>
    </div>
    <button
      aria-label="Edit user name"
      class="user-header_icon__44nqd43 style_transition_all__192bcwg41 style_enabledIconButtonVariants_teal__92u52zz style_enabledIconBase__92u52z5 style_iconBase__92u52z1"
    >
      <span
        class="pencil_medium_teal__2ixuvsb style_iconMedium__p90ga82 style_iconBase__p90ga80"
      />
    </button>
  </div>
</div>
`;
